### PR TITLE
Add resource quotas and limit ranges.

### PR DESCRIPTION
**Pull request by moolitayer**
Adding resource quotas & limit ranges.

### CODE REVIEW

**Comment by moolitayer** [comment]
fyi @simon3z @abonas

File: README.md [hunk 1]
```diff
@@ -118,7 +118,7 @@ client = Kubeclient::Client.new 'https://localhost:8443/api/' , 'v1beta3',
 ## Examples:
 
 #### Get all instances of a specific entity type
-Such as: `get_pods`, `get_secrets`, `get_services`, `get_nodes`, `get_replication_controllers`
+Such as: `get_pods`, `get_secrets`, `get_services`, `get_nodes`, `get_replication_controllers`, `get_resource_quotas`, `get_limit_range`
```
Comment (abonas) [inline README.md]: why is this not plural?

**Comment by abonas** [comment]
@moolitayer this PR's title is too generic. Please make it a clear one.

File: lib/kubeclient/inflections.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+ActiveSupport::Inflector.inflections do |inflect|
+  inflect.irregular 'resource_quota', 'resource_quotas'
+  inflect.irregular 'resourcequota', 'resourcequotas'
+end
```
Comment (abonas) [inline lib/kubeclient/inflections.rb]: @tenderlove @matthewd it seems that the inflector doesn't pluralize the word quota.
I saw it also here
http://codereply.com/answer/8hnvfd/method-pluralize-seems-working.html
I wanted to make sure that the above solution is the recommended one.
(Also based on what's written here https://github.com/rails/rails/blob/5373bf228d1273deae0ed03370ec4a63c580422b/activesupport/lib/active_support/inflector/methods.rb#L11)

File: test/test_limit_range.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  def test_get_from_json_v3
+    stub_request(:get, %r{/limitranges})
+      .to_return(body: open_test_json_file('limit_range.json'),
+                 status: 200)
+
+    client = Kubeclient::Client.new 'http://localhost:8080/api/', 'v1'
+    limits = client.get_limit_range 'limits', 'quota-example'
+
+    assert_instance_of(Kubeclient::LimitRange, limits)
+    assert_equal('limits', limits.metadata.name)
```
Comment (abonas) [inline test/test_limit_range.rb]: nit, but the name is a bit confusing - you are pulling a single entity here

File: test/test_limit_range.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    assert_equal('Container', limits.spec.limits[0].type)
+
```
Comment (abonas) [inline test/test_limit_range.rb]: no validation for the content of the limit?

File: lib/kubeclient/inflections.rb (hunk 2 above)
Comment (matthewd) [inline lib/kubeclient/inflections.rb]: I *think* you can just do `inflect.irregular 'quota', 'quotas'`, and it'll handle it as a suffix.

But either way, yes, this looks right.

The only potential problem, is that the library is now registering custom pluralizations which could affect the surrounding application -- that's unlikely to be something a developer expects when adding your gem to their Gemfile. I don't know of any better solution for that, though.

Comment (abonas) [inline lib/kubeclient/inflections.rb]: > The only potential problem, is that the library is now registering custom pluralizations which could affect the surrounding application -- that's unlikely to be something a developer expects when adding your gem to their Gemfile. I don't know of any better solution for that, though.

so hardcoding it without messing with the inflector would be better?

Comment (matthewd) [inline lib/kubeclient/inflections.rb]: Yeah, if there's a way you can do that without uglifying your code too much, that's probably a more "pure" choice, for a library.

That said, in practice, I doubt you'd actually upset/inconvenience anyone by defining the inflection. ¯\\\_(ツ)_/¯

Comment (abonas) [inline lib/kubeclient/inflections.rb]: > Yeah, if there's a way you can do that without uglifying your code too much, that's probably a more "pure" choice, for a library.

oh it surely will uglify :)
because it must add an "if"

File: README.md (hunk 1 above)
Comment (moolitayer) [inline README.md]: done.

File: test/test_limit_range.rb (hunk 4 above)
Comment (moolitayer) [inline test/test_limit_range.rb]: fixed

File: test/test_limit_range.rb (hunk 3 above)
Comment (moolitayer) [inline test/test_limit_range.rb]: fixed to limit_range

File: README.md (hunk 1 above)
Comment (moolitayer) [inline README.md]: fixed.

File: test/test_limit_range.rb (hunk 4 above)
Comment (moolitayer) [inline test/test_limit_range.rb]: added validations.

**Comment by moolitayer** [comment]
@abonas 
>> @moolitayer this PR's title is too generic. Please make it a clear one.
what would you like to see?

File: lib/kubeclient/inflections.rb (hunk 2 above)
Comment (moolitayer) [inline lib/kubeclient/inflections.rb]: @matthewd 
>> I think you can just do inflect.irregular 'quota', 'quotas', and it'll handle it as a suffix.
That works. thanks

@abonas it seems that this is the preferred solution (Inflector)?

**Comment by abonas** [comment]
> what would you like to see?

for example, *which* entities are added in the patch

File: lib/kubeclient/inflections.rb (hunk 2 above)
Comment (abonas) [inline lib/kubeclient/inflections.rb]: > @abonas it seems that this is the preferred solution (Inflector)?

@moolitayer yes
@matthewd thanks a lot for your inputs!!!

File: lib/kubeclient/inflections.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+ActiveSupport::Inflector.inflections do |inflect|
+  inflect.irregular 'quota', 'quotas'
```
Comment (simon3z) [inline lib/kubeclient/inflections.rb]: Is this global? Any way to scope this just for this gem? Maybe @matthewd or @tenderlove know that.

**Comment by simon3z** [comment]
One question about the Inflector (inline), the rest looks OK :+1:

File: lib/kubeclient/inflections.rb (hunk 5 above)
Comment (matthewd) [inline lib/kubeclient/inflections.rb]: [Nope](https://github.com/abonas/kubeclient/pull/106#discussion-diff-36607177)

**Comment by simon3z** [comment]
The `active_support` inflection culprit is:

    inflect.plural(/([ti])a$/i, '\1a')

https://github.com/rails/rails/blob/3-2-stable/activesupport/lib/active_support/inflections.rb#L14

For the same reason mentioned here:

https://github.com/rails/rails/blob/5373bf228d1273deae0ed03370ec4a63c580422b/activesupport/lib/active_support/inflector/methods.rb#L11

>  # The Rails core team has stated patches for the inflections library will not
>  # be accepted in order to avoid breaking legacy applications which may be
>  # relying on errant inflections.

I'd prefer to have a local wrapper here just in our gem (to not affect interaction with other libraries), something like:

    def pluralize_entity(entity_name)
        return 'quotas' if entity_name == 'quota'
        entity_name.pluralize
    end

and then we'll just call our `pluralize_entity`.

**Comment by moolitayer** [comment]
@abonas agreement on a local wrapper?

**Comment by abonas** [comment]
> @abonas agreement on a local wrapper?

@simon3z @moolitayer 
I'm not a big fan of, plus @matthewd said it's not *that* bad to configure the inflector, but if you tend strongly towards this local solution, I won't block it.

**Comment by moolitayer** [comment]
@simon3z @abonas 
In my opinion, as I've expressed to Alissa, the best solution would be to replace
```
%w(Pod Service ReplicationController Node Event Endpoint Namespace Secret ResourceQuota LimitRange)
```
with:
```
[['pod', 'pods'], ['Service', 'Services'] ...
```

Yes its more verbose but feels better for api end points then a text / localization solution with multiple pluralize/singularize calls.

Any way since we are 3 people with 3 different opinions any chance we agree to go with the current solution and patch it up to what ever solution in the future if need arises ?
(also I'm lazy)

**Comment by jimmidyson** [comment]
Just noticed - isn't it `Endpoints`, not `Endpoint`? Type is `endpoints` - https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1/types.go#L1249

**Comment by simon3z** [comment]
> Just noticed - isn't it Endpoints, not Endpoint? Type is endpoints -

Where exactly? we keep all the entities names singular and then we pluralize them when required.

     ENTITY_TYPES = %w(Pod Service ReplicationController Node Event Endpoint
                       Namespace Secret).map do |et|

**Comment by jimmidyson** [comment]
@simon3z Which works for most types, but `Endpoints` is an exception: it's pluralized in the Kubernetes v1 API kind as linked above.

**Comment by abonas** [comment]
> @simon3z Which works for most types, but Endpoints is an exception: it's pluralized in the Kubernetes v1 API kind as linked above.

@simon3z @jimmidyson known issue
https://github.com/abonas/kubeclient/issues/81

@moolitayer, I'm not in favor of copy-pasting all entities, so it's either the inflector configuration or the solution suggested by @simon3z . Both work for me.

**Comment by simon3z** [comment]
> In my opinion, as I've expressed to Alissa, the best solution would be to replace with:
> 
>     [['pod', 'pods'], ['Service', 'Services'] ...
> 
> Yes its more verbose but feels better for api end points then a text / localization solution with multiple pluralize/singularize calls.

Fine by me if you prefer it and you think that code would be nicer than my proposal :+1:
(I have the feeling that you'll have to deal with uppercase/lowercase)

> I'm not a big fan of, plus @matthewd said it's not that bad to configure the inflector, but if you tend strongly towards this local solution, I won't block it.

Touching a singleton that is shared between multiple libraries is a big no-no for me. (Think how upset you would be if one of the other libraries would touch one of our pluralizations :smile:  )

**Comment by simon3z** [comment]
I haven't invested much time in reviewing documentation and tests, we'd need another couples of eyes there. The core LGTM :+1:

**Comment by moolitayer** [comment]
@simon3z @abonas 
1) fyi afaiu releasing a version of kubeclient might cause problems;
#92 removed the ssl_options method from common.rb which is used by openshift_client.
the dependency on kubeclient at openshift_client gemspec is stated as: 
spec.add_dependency 'kubeclient', '>= 0.1.17'
but we don't want old openshift_clients to get new kubeclient and break.
afaiu this means we should release kubeclient not before we have a new version of openshift_client.

2) after making the fix to openshift_clien I have not yet found a way to test it - I was unable to force openshift_client work with the new version of kubeclient using it's gemspec (like you can do :path => 'local/' in a gem file) so I am stuck. please advise

**Comment by abonas** [comment]
> after making the fix to openshift_clien I have not yet found a way to test it - I was unable to force openshift_client work with the new version of kubeclient using it's gemspec (like you can do :path => 'local/' in a gem file) so I am stuck. please advise

have you tried bumping kubeclient dep in gemspec to 0.2.0 or 0.3.0 in openshift_gem?
Why do you need to use a local path to test it? 
https://github.com/abonas/kubeclient/pull/92 was released in one of the above releases, iirc 0.2.0

In any case, the way I know to test a local gem is as following (only for local tests, not to be commited): 
1. create a Gemfile in openshift client. 
2. Add there the dependency you want to kubeclient with the local path (gem kubeclient .....) . 
3. Add there also a reference to the gemspec (iirc just write there "gemspec"), 
4. remove the version requirement for kubeclient in gemspec.

**Comment by simon3z** [comment]
@abonas LGTM :+1: ready for merge

File: README.md [hunk 6]
```diff
@@ -177,9 +177,9 @@ client.create_service service`
 ```
 
 #### Update an entity
-For example: `update_pod`, `update_service`, `update_replication_controller`, `update_secret`
+For example: `update_pod`, `update_service`, `update_replication_controller`, `update_secret`, `update_resource_quota`, `update_limit_range`
 
-Input parameter - object of type `Service`, `Pod`, `ReplicationController`
+Input parameter - object of type `Pod`, `Service`, `ReplicationController` etc.
```
Comment (abonas) [inline README.md]: why is this change relevant to the patch and what is its purpose? (just swaping pod and service in places?)

Comment (moolitayer) [inline README.md]: to align with the text above: `update_pod`, `update_service`.
do you want this removed?

Comment (abonas) [inline README.md]: > to align with the text above: update_pod, update_service.
> do you want this removed?

in the future please separate changes like this to a separate patch since it is unrelated and it confuses when reviewing.
but nonetheless such alignment fixes are welcome.
